Enhance: show profile indication in central debug console output

SlySven opened the pull request:
This PR adds a single character inside `[`...`]` to the beginning of (nearly) all texts displayed on the Central Debug Console which is made unique for each session of a profile run in a Mudlet application run. This marking (except for messages not from a particular profile) is not displayed when there is only one profile - and the start and end (and a change of name) of each profile is noted and a revised listing of the symbol used for each active profile is shown each time it changes.

To do this the previous dummy `>> 0` used at the end of a `TDebug()` message creation has been changed to instead take a pointer to a profile's `host` instance - from which the profile name is extracted. To allow for multi-coloured messages to be concatenated, inserting a `(QChar) TDebug::csmContinue` element at the start of a new `TDebug` text will cause it to be appended onto the end of a previous `TDebug` text without acquiring the profile indication marking.

Messages that are generated with a `nullptr` as the Host pointer are expected to be messages from the Mudlet Application rather than a particular profile and they will always be marked with a `[*]`.

When multi-playing closing a profile and reopening it again will cause a new identifier to the used for the subsequent run - so it can be seen which run any messages are from.

The current design uses identifiers from the sequence: '1' to '9', then 'A' to 'Z', then 'a' to 'z' - after that set of 61 marks are exhausted then all subsqeuent profile runs in that Mudlet application run are shown with a '?'. it would be possible to extend the sequence but as the Central Debug console is intended to help with debugging it is not expected for the finite number of identifiers to be a serious limitation!

This will close #804 (originally a LaunchPad issue: https://bugs.launchpad.net/mudlet/+bug/1658271 )!

Also:
* simplify some Lua handling of boolean returns and conditions based upon an integer value.
* in `TTrigger::match(...)` use `QStringList::at(...)` in some places where it is faster/safer than `QStringList::operator[]()`.
* Remove a rogue line-feed prior to a: "LUA: code compiled without errors. OK" message which was making it more spammy than it needed to be.


https://user-images.githubusercontent.com/110988/111956028-944d3400-8aea-11eb-9dc1-220e6c224383.mp4


([streamable.com link](https://streamable.com/mnrv9e))



Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: show profile indication in central debug console output

This PR adds a single character inside `[`...`]` to the beginning of
(nearly) all texts displayed on the Central Debug Console which is made
unique for each session of a profile run in a Mudlet application run.
This marking (except for messages not from a particular profile) is not
displayed when there is only one profile - and the start and end (and
a change of name) of each profile is noted and a revised listing of the
symbol used for each active profile is shown each time it changes.

To do this the previous dummy `>> 0` used at the end of a `TDebug()`
message creation has been changed to instead take a pointer to a profile's
`host` instance - from which the profile name is extracted. To allow for
multi-coloured messages to be concatenated, inserting a
`(QChar) TDebug::csmContinue` element at the start of a new `TDebug` text
will cause it to be appended onto the end of a previous `TDebug` text
without acquiring the profile indication marking.

Messages that are generated with a `nullptr` as the Host pointer are
expected to be messages from the Mudlet Application rather than a
particular profile and they will always be marked with a `[*]`.

When multi-playing closing a profile and reopening it again will cause a
new identifier to the used for the subsequent run - so it can be seen
which run any messages are from.

The current design uses identifiers from the sequence:
'1' to '9', then 'A' to 'Z', then 'a' to 'z' - after that set of 61 marks
are exhausted then all subsqeuent profile runs in that Mudlet application
run are shown with a '?'. it would be possible to extend the sequence but
as the Central Debug console is intended to help with debugging it is not
expected for the finite number of identifiers to be a serious limitation!

This will close #804 (originally a LaunchPad issue:
https://bugs.launchpad.net/mudlet/+bug/1658271 )!

Also:
* simplify some Lua handling of boolean returns and conditions based upon
an integer value.
* in `TTrigger::match(...)` use `QStringList::at(...)` in some places where
it is faster/safer than `QStringList::operator[]()`.
* Remove a rogue line-feed prior to a:
"LUA: code compiled without errors. OK"
message which was making it more spammy than it needed to be.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
![image](https://user-images.githubusercontent.com/6163092/107739967-f6539600-6d01-11eb-883b-ae9374b4cfcc.png)

![image](https://user-images.githubusercontent.com/6163092/107740471-1b94d400-6d03-11eb-9821-38a7beee869f.png)

SlySven committed:
Revise: fix something that was disabled in my test environment

I was using a Qt version that did not have the TTS environment enabled so
some code was `#ifdef`fed out - and fa9led to work when it wasn't excluded!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: keep CodeFactor happy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented on src/TConsole.cpp:
```diff
@@ -1575,17 +1574,17 @@ int TConsole::select(const QString& text, int numOfMatch)
 {
     if (mudlet::debugMode) {
-        TDebug(QColor(Qt::darkMagenta), QColor(Qt::black)) << "\nselectSection(" << from << "," << to << "): line under current user cursor: " << buffer.line(mUserCursor.y()) << "\n" >> 0;
+        TDebug(Qt::darkMagenta, Qt::black) << "selectSection(" << from << "," << to << "): line under current user cursor: \n\"" << buffer.line(mUserCursor.y()) << "\"\n" >> mpHost;
```
How about single quotes instead? 

```suggestion
        TDebug(Qt::darkMagenta, Qt::black) << "selectSection(" << from << "," << to << "): line under current user cursor: \n'" << buffer.line(mUserCursor.y()) << "'\n" >> mpHost;
```

SlySven replied:
Actually I had gone through and revised a whole load of messages - then I realised I was yielding to *creature feep* and undid them all - apart from this and one other that I didn't spot! I will in fact put that part back and leave all of that to some future PR - and that *is* needed IMVHO...!

Kebap replied:
Even better decision!

Kebap commented on src/TDebug.cpp:
```diff
@@ -25,22 +25,67 @@
-// of operator<<() calls) causes the accumulated message to be printed/flushed
-// out...
-TDebug& TDebug::operator>>(const int code)
+// This is the method that pushes the accumulated text out to the Central Debug
+// Console. If the accumulated message text (in msg) begins with csmContinue
+// then this message is appended onto a previous one and should NOT include
+// any indicator (at the beginning) of the profile from which it came - which
+// is now deduced from the supplied Host pointer.
+TDebug& TDebug::operator>>(Host* pHost)
```
This in part repeats a lengthy comment from 10 lines prior, please separate causes or pull together in one.

SlySven replied:
:mag_right: The comments do server different purposes but I agree there is some redundancy in this text. I will see what I can eliminate...

Kebap commented on src/TDebug.cpp:
```diff
@@ -122,9 +156,117 @@ TDebug& TDebug::operator<<(const QList<QString>& list)
+        TDebug::smIdentifierMap.insert(pHost, newIdentifier);
+    }
+    TDebug localMessage(Qt::blue, Qt::white);
+    localMessage << QStringLiteral("Profile: '%1' started.\n").arg(hostName) >> nullptr;
```
```suggestion
    localMessage << QStringLiteral("Profile '%1' started.\n").arg(hostName) >> nullptr;
```
Remove punctuation in sentence middle

SlySven replied:
:+1: :-1: This is a bit subjective IMHO - but I *could* live without the ':' I suppose...

Kebap replied:
Please do, or otherwise show sentences like "street: 'sesame' built." It's just wrong grammar, no subjectivity required.

SlySven replied:
*I beg to differ **:¹*** https://christopheraltman.wordpress.com/tag/colon-basics-punctuation-toolbox/ is one person's attempt to introduce it as an *introducer*...

However I am doing it in the next commit,

***¹** - Due to some oddity in the GH markdown I have to put a space before that `:` otherwise it won't enable me to embolden it and not the preceding characters!*

Kebap replied:
I like to discuss grammar, and the examples you show as introducing are always preceded by what could otherwise stand as a sentence of its own. I would agree if your sentence read "street built: 'sesame'" but not as it was. Thanks for removing that then.

Kebap replied:
For comparison: https://writingcenter.unc.edu/tips-and-tools/semi-colons-colons-and-dashes/#:~:text=How%20to%20check%20for%20mistakes

> How to check for mistakes
Ask yourself a question: does the material preceding the colon stand on its own? One way to tell if the colon has been properly used is to look only at the words that come in front of the colon. Do they make a complete thought? If not, you may be using the colon improperly. Check above to see if you have made one of the most common mistakes.

Where number 1 was "Using a colon between a verb and its object"

Source: University of North Carolina

Kebap commented on src/TDebug.cpp:
```diff
@@ -122,9 +156,117 @@ TDebug& TDebug::operator<<(const QList<QString>& list)
+{
+    TDebug::smIdentifierMap.remove(pHost);
+    TDebug localMessage(Qt::darkGray, Qt::white);
+    localMessage << QStringLiteral("Profile: '%1' ended.\n").arg(pHost->getName()) >> nullptr;
```
```suggestion
    localMessage << QStringLiteral("Profile '%1' ended.\n").arg(pHost->getName()) >> nullptr;
```
Remove punctuation in sentence middle

SlySven replied:
As above - so below.

Kebap commented on src/TDebug.cpp:
```diff
@@ -122,9 +156,117 @@ TDebug& TDebug::operator<<(const QList<QString>& list)
+        messageLines.prepend(QStringLiteral("There are %1 profiles active now, so each message from a profile will be "
+                                            "prefixed with a single character inside a pair of '['...']'s as follows:").arg(TDebug::smIdentifierMap.count()));
+    } else {
+        messageLines.prepend(QStringLiteral("There is only a single profile active right now, so only messages not from that "
+                                            "profile will be prefixed though the indicated character will be retained and "
+                                            "used should another profile be opened:"));
```
While smIdentifierMap.count() <= 1, do not print these messages and profile names at all, yet.

SlySven replied:
Not sure I get you? :confused: 

As a log/recording (possibly by review by someone else) of a run of Mudlet the information is needed IMO even when only **one** profile is active.

vadi2 replied:
I agree, for just one not necessary - the problem you mention isn't something we've faced from experience. Let's solve it when we do

Kebap replied:
My comment from above was relevant here for this, so copy it here:

> Why do the Alias messages in your first screenshot wear no prefix? Your message seems to indicate [1] instead.
DO we want to differentiate system messages from profile messages also if only one profile is active as well - or not?
If not, drop the message, as it mostly creates confusion and not clarifies anything before other profiles are added.

Graphical overview:
![image](https://user-images.githubusercontent.com/117238/107806583-56305800-6d67-11eb-8705-3559aed133e0.png)

The lines "profile started" and "profile ended" can remain, but the others should vanish.

SlySven replied:
However - in the regions that you have said that the marking are not required - "system" messages (or if a developer  uses the old form and ends a message with `>> 0;` AND the compile swallows that - I am not sure if they do or not - perhaps I should check!) will still be marked with a `[*]` whereas the profile ones will not be marked - but showing the mark *that would be used* was intended to help keep track when things are changed - perhaps the revised messages in the next commit will put it better...

Kebap replied:
That is why I asked (and copy one more time):

> DO we want to differentiate system messages from profile messages also if only one profile is active as well - or not?

We did not settle that, yet.
Only after we decide that, can we do away with the [*] or keep it like during multiplay.

> help keep track when things are changed

Hence the "profile started/ended" remain.

SlySven replied:
There is code in place that allows for an active Host to be renamed (not sure if it is utilised BTW) - that also will tickle this code but it is handled by this PR.

Obviously I am biased towards what I have done but to my mind it does not make sense to only mark non-profile specific messages when more than one profile is running - as they are still not specific to a profile when only one is active. Whilst currently that is only applicable to the messages notifying of changes there is nothing, even in the original code, to say that `TDebug` messages are only for the profile. OTOH we still have the limitation that one cannot ***start*** to show the CDC without having a profile at least loaded - which is another issue I want to address (eventually) - as it  makes it harder to debug profile loading problems than it needs to be.

Kebap replied:
> it does not make sense to only mark non-profile specific messages when more than one profile is running

Just they were not differentiated in Mudlet prior. 
Only came about because you want to indicate profile number (good) when multiple are present and they are in none. 
When there is but one profile loaded, they can just as well remain undifferentiated as they were until now.

SlySven replied:
> Just they were not differentiated in Mudlet prior.

Yeah, because there wasn't any like that before - but there are now and there could be more in the future...

Kebap replied:
Let's cross the bridge when we get there. 
Until then, remove these 4 lines from debug until players actually start to multiplay.
➡️ YAGNI
https://medium.com/better-programming/yagni-you-aint-gonna-need-it-f9a178cd8e1
![](https://imgs.xkcd.com/comics/the_general_problem.png)

Kebap commented on src/TDebug.cpp:
```diff
@@ -122,9 +156,117 @@ TDebug& TDebug::operator<<(const QList<QString>& list)
+    }
+    messageLines.prepend(QStringLiteral(" [*] = System message, not belonging to a specific profile"));
+    if (TDebug::smIdentifierMap.count() > 1) {
+        messageLines.prepend(QStringLiteral("There are %1 profiles active now, so each message from a profile will be "
+                                            "prefixed with a single character inside a pair of '['...']'s as follows:").arg(TDebug::smIdentifierMap.count()));
```
```suggestion
        messageLines.prepend(QStringLiteral("Now %1 profiles are active. Each message will be prefixed as follows:")
        .arg(TDebug::smIdentifierMap.count()));
```

Shorten message, drop obvious information.

SlySven replied:
Guess *some* improvements could be made here.

Kebap commented on src/TDebug.h:
```diff
@@ -27,31 +27,82 @@
+    TDebug& operator<<(const QList<QString>&);
+    TDebug& operator<<(const QList<int>&);
+
+    // one of two Unicode BMP non-characters, prepend this to any continuation
+    // message to suppress the insertion of the profile identifying marking.
+    static const QChar csmContinue;
```
Again the same comment information from the cpp file is duplicated here

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -11247,35 +11249,6 @@ QString TLuaInterpreter::formatLuaCode(const QString &code)
     return result;
 }
 
-// No documentation available in wiki - internal function
-bool TLuaInterpreter::compileScript(const QString& code)
-{
-    lua_State* L = pGlobalLua;
-
-    int error = luaL_dostring(L, code.toUtf8().constData());
-    if (error != 0) {
-        std::string e = "no error message available from Lua";
-        if (lua_isstring(L, 1)) {
-            e = "Lua error:";
-            e += lua_tostring(L, 1);
-        }
-        if (mudlet::debugMode) {
-            TDebug(QColor(Qt::white), QColor(Qt::red)) << "LUA: code did not compile: ERROR:" << e.c_str() << "\n" >> 0;
-        }
-    } else {
-        if (mudlet::debugMode) {
-            TDebug(QColor(Qt::white), QColor(Qt::darkGreen)) << "LUA: code compiled without errors. OK\n" >> 0;
-        }
-    }
-    lua_pop(L, lua_gettop(L));
-
-    if (error == 0) {
-        return true;
-    } else {
-        return false;
-    }
-}
-
```
You remove this whole function and don't even mention it in PR description. Why?

SlySven replied:
I found it was unused and unneeded - but I forgot to document it in the commit message.

***Note to self: include something about this in the* Squash-and-merge *message.***

Kebap replied:
Would not expect such change under a title "show profile indication in debug" at all - would you?

SlySven replied:
Yeah, but it has `TDebug` messages that otherwise I'd have to revise - and why revise when one can nuke? :smiling_imp:

Kebap commented on src/TTrigger.cpp:
```diff
@@ -215,16 +216,16 @@ bool TTrigger::setRegexCodeList(QStringList regexList, QList<int> propertyList)
-                    TDebug(QColor(Qt::white), QColor(Qt::red)) << "REGEX ERROR: failed to compile, reason:\n" << error << "\n" >> 0;
-                    TDebug(QColor(Qt::red), QColor(Qt::gray)) << R"(in: ")" << regexp.constData() << "\"\n" >> 0;
+                    TDebug(Qt::white, Qt::red) << "REGEX ERROR: failed to compile, reason:\n" << error << "\n" >> mpHost;
+                    TDebug(Qt::red, Qt::white) << TDebug::csmContinue << R"(in: ")" << regexp.constData() << "\"\n" >> mpHost;
```
this file contains a bunch of unexpected color changes here, not just removed QColor

SlySven replied:
Another one of the changes I didn't undo. As it happens I found that "red-on-grey" is visually (for me who does have normal colour perception) a poor choice - but will revert for another PR.

Note that all the explicit `QColor(`...`)` wrappers in the `TDebug` constructor uses are redundant and since I was visiting every one of them to put the `Host*` on the end it made sense to remove them at the same time.

Kebap replied:
> will revert for another PR

Good call

> made sense to remove them

Applaud that effort

Kebap replied:
You undid this very instance but there were a handful. Please review

SlySven replied:
Done, I think.

Kebap commented on src/mudlet.cpp:
```diff
@@ -266,6 +266,7 @@ mudlet::mudlet()
     menuHelp->setToolTipsVisible(true);
     menuAbout->setToolTipsVisible(true);
 
+    // FIXME: add a command line option to change this?
```
Maybe open a public issue here instead of adding TODO comments to source?

SlySven replied:
Initially a coding note to myself - remember I use the **ToDo** plugin so it is helpful to *me* - but will make that issue before I merge this PR and link to it from this comment.

SlySven replied:
Issue created #4807 .

Kebap reviewed (commented):
I like the idea to show a prefix indicating the originating profile here!
Thanks for the screenshots! Helps to anticipate the changes. I did not test them myself yet.

Why do the Alias messages in your first screenshot wear no prefix? Your message seems to indicate [1] instead.
DO we want to differentiate system messages from profile messages also if only one profile is active as well - or not?
If not, drop the message, as it mostly creates confusion and not clarifies anything before other profiles are added.

> after that set of 61 marks are exhausted then all subsequent profile runs in that Mudlet application run are shown with a '?'. it would be possible to extend the sequence but as the Central Debug console is intended to help with debugging it is not expected for the finite number of identifiers to be a serious limitation!

Please just restart the sequence, instead of continuing with question marks only.

SlySven commented on src/TConsole.cpp:
```diff
@@ -1546,9 +1546,8 @@ int TConsole::select(const QString& text, int numOfMatch)
     }
 
     if (mudlet::debugMode) {
-        TDebug(QColor(Qt::darkMagenta), QColor(Qt::black)) << "\nline under current user cursor: " >> 0;
```
Can't retain the leading `\n` as that won't sit well (IMHO) with something prefixed onto the line.

SlySven committed:
Revise: changes suggested by peer-review

Note that `(bool) TLuaInterpreter::compileScript(const QString&)` has been
found to be unused so has been removed from the codebase.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Why do the Alias messages in your first screenshot wear no prefix? Your message seems to indicate [1] instead.

Because I only show them if needed and if there is ***only one*** profile active it isn't needed - however the `[x]` is still shown in the tables so the viewer can relate the display before and after a change in the profiles that are active.

> > after that set of 61 marks are exhausted then all subsequent profile runs in that Mudlet application run are shown with a '?'. it would be possible to extend the sequence but as the Central Debug console is intended to help with debugging it is not expected for the finite number of identifiers to be a serious limitation!
>
> Please just restart the sequence, instead of continuing with question marks only.

I am not sure I can do the logic for that with the code I have. Note that *resetting* a profile (with `resetProfile()`) does not cause the `Host` - or the tag - to change - so this will only became an issue for someone who loads and unloads 62 profiles in a single Mudlet run. That seems to be a rather unlikely corner case and one I'd rather not have to worry about until a proven need comes to light...! :worried: 

Humm, it seems I wrote this comment several hours ago but never posted it.

Kebap commented:
> I am not sure I can do the logic for that with the code I have. 
> this will only became an issue for someone who loads and unloads 62 profiles in a single Mudlet run. That seems to be a rather unlikely corner case and one I'd rather not have to worry about until a proven need comes to light...! 😟

Yep, keep Mudlet running for 2-3 days and open and close a few muliplay profiles - does not seem so unlikely to me really. Then their debug window spews only [?] anymore.
The logic seems simple with traversing the list modulo style, making sure not to pick symbols that are still in use for other hosts.
Then again, you used my above argument against me, and may postpone until players complain. Just be ready if they do.

vadi2 commented:
I agree we should not fall back to [[?]]. To be honest the #'s in the console window should reflect the profile tabs' index to me - that seems most intuitive. They should not rotate in their own order: that will be confusing.

SlySven commented:
> To be honest the #'s in the console window should reflect the profile tabs' index to me - that seems most intuitive. They should not rotate in their own order: that will be confusing.

That will not be constant (we do want to be able to rearrange them don't we, remember #4790 is only a few days into its ex-PR state!) and will likely change as other profiles are opened/closed anyhow.

vadi2 commented:
Yes, but I believe that will be most intuitive / simple.

Kebap commented:
> the #'s in the console window should reflect the profile tabs' index to me - that seems most intuitive. They should not rotate in their own order: that will be confusing.

If so we would need (not only "profile started" and "profile ended") messages to indicate "profile rearranged" announcing the new world order from that point in time on

SlySven commented:
~~Only the first 9 profile instances will be using a number - and to perhaps avoid confusing them with the *tab* number you are proposing elsewhere I should remove the digits `1` to `9` from the potential tags...~~


*Edit: (update)*... done now in https://github.com/Mudlet/Mudlet/pull/4781/commits/bb419fd2402dc6fb00cc34aa0a065eefe45fc1b1 - only upper case A to Z are used - should some wierdo have more than 26 profiles open ***at a time*** then the excess will get `[?]` but less than that pathalogical case each new profile opened will use the remaining letters out of `A` to `Z` initially and then oldest one that was closed after 26 have been opened and closed.

SlySven committed:
Revise: do the allocation of tags differently

Reuse tags by enqueuing tags as they are finished with so as to be able to
be available if all previously unused ones get used. This drops the number
of tags down to just the capital letters of the alphabet - but this will
support 26 profiles running at a time and to be perfectly honest - that
seems more than enough. Should some weirdo try and run more than that then
the additional profiles will be identified with a `[?]`...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_showProfileInDebugOutput

Bring in recent fixes in development branch.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Delwing commented on src/TKey.cpp:
```diff
@@ -158,7 +159,7 @@ void TKey::compile()
         if (!compileScript()) {
             if (mudlet::debugMode) {
-                TDebug(Qt::white, Qt::red) << "ERROR: Lua compile error. compiling script of key binding:" << mName << "\n" >> 0;
+                TDebug(Qt::white, Qt::red) << "ERROR: Lua compile error. compiling script of key binding:" << mName << "\n" >> mpHost;
```
If re require host to be added to debug message, shouldn't it be enforced by param? This way we have to know convention and "remember".

SlySven replied:
I am not entirely sure whether a literal `0` in the source code might not be capable of being converted to a `nullptr` - I think not in modern C++17 code (as part of the purpose of `nullptr` is to avoid it being mistaken as something else) and I *think* the compiler will start to moan at you if you do not change that `0`; but, as someone who comes from a C coding background, I am aware that C89, for instance, is a bit more lax about this sort of thing...

SlySven replied:
The original proposal paper might clarify things: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2431.pdf

Delwing replied:
Ok, wasn't sure so thought that it is better to ask than keep quiet :)

SlySven committed:
Merge branch 'development' into Enhance_showProfileInDebugOutput

SlySven committed:
BugFix: fix typo in previous merge commit

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TDebug.cpp:
```diff
@@ -25,22 +25,65 @@
+                if (message.mTag.isNull()) {
+                    mudlet::mpDebugConsole->print(message.mMessage, message.mForeground, message.mBackground);
+                } else {
+                    mudlet::mpDebugConsole->print(QStringLiteral("[%1] %2").arg(message.mTag, message.mMessage), message.mForeground, message.mBackground);
```
Not being able to ***move*** construct the return value because it is generated from a `const` expression (a `QStringLiteral` with parameter replacement) is going to be a pattern that we experience a lot I suspect.

SlySven committed:
Enhance: add tags to tab bar as well as CDC

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> https://github.com/Mudlet/Mudlet/pull/4781#issuecomment-778821293 

As #4976 will enable the situation that was raised it is just as well that I am not longer using numbers... :grinning:

SlySven committed:
Merge branch 'development' into Enhance_showProfileInDebugOutput

Needed to acquire new functionalility related to moving tabs in tab bar
and their associated profiles.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: use a different symbol (heavy asterix) instead of a plain one

This is for non-profile specific messages and sits better on the text line
and is bigger/heavier than a standard (ASCII) Asterix symbol.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Looking into reviewing it now. Sorry for the delay. You've done a lot of work on it already - so I'll send a PR instead of piling more comments.

vadi2 commented:
A PR with improvements to get this moving along - @SlySven @Kebap check out https://github.com/SlySven/Mudlet/pull/24.

Kebap commented:
> A PR with improvements to get this moving along - @SlySven @Kebap check out [SlySven#24](https://github.com/SlySven/Mudlet/pull/24).

Very nice, I like!

vadi2 committed:
Enhance show profile in debug output (#24)

* Don't show message for just 1 profile

* Fix bug: closing debug via X didn't remove indicators

We really should be emitting a proper Mudlet event for this!

* Don't show a tag if we can't tell where is it from

* Update src/mudlet.cpp

SlySven committed:
BugFix: correct error in Vadi's last squashed down commits to his (#24) PR

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Uuugh! The # 24 in those titles is actually https://github.com/SlySven/Mudlet/pull/24 !

Anyhow - I am still not convinced that removing the listing when the penultimate profile is close is that good an idea, consider this:
![image](https://user-images.githubusercontent.com/6163092/112089192-0ffdbe00-8b89-11eb-98cc-dc36314da9b4.png)

That had three profiles open until the third `[C] WoTMUD_Jath` one was closed, and the remaining two were reidentified, then the second remaining one `[D] Testing` (I'd already opened and closed one) was closed - leaving it not quite so obvious that the remaining one `[A]` was left.

If, as I believe some might wish to in some cases, we actually write the CDC to a log file, then to confirm the profile that was still being shown in the debug output one would have to ***scroll back*** to the previous change (when `WoTMUD_Jath` was closed - which in this case was just prior but need not be at all - there could be hundreds of lines between them) to confirm that with `[D]` having been closed, that would leave `[A]` - which would not be shown when we went from 2 to 1 profiles.

SlySven committed:
Revise: keep CodeFactor happy and update Vadi's copyright for prior commit

CodeFactor wanted the TTabBar constructor with a single `Host*` argument
to be marked `explicit`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> leaving it not quite so obvious that the remaining one [A] was left.

I don't think it'll be a serious problem in practice. Let's give it a go with this, and if I'm incorrect we'll adjust it :+1: 

We've taken similar gambles recently, for example with the map info - which people don't have complaints with - so the process is working okay.
